addons/l10n_tw_edi_ecpay/tests/test_edi.py

```diff
@@ -0,0 +1,338 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from datetime import datetime
+from freezegun import freeze_time
+from urllib.parse import urljoin
+
+from odoo.addons.account.tests.test_account_move_send import TestAccountMoveSendCommon
+from odoo.addons.l10n_tw_edi_ecpay.controllers.main import EcpayInvoiceController
+from odoo import http
+from odoo.exceptions import UserError
+from odoo.tests import tagged
+from odoo.tests.common import HttpCase
+from unittest.mock import patch
+
+CALL_API_METHOD = 'odoo.addons.l10n_tw_edi_ecpay.utils.EcPayAPI.call_ecpay_api'
+
+
+@tagged('post_install_l10n', 'post_install', '-at_install')
+class L10nTWITestEdi(TestAccountMoveSendCommon, HttpCase):
+    def setUp(self):
+        super().setUp()
+        self.EcpayInvoiceController = EcpayInvoiceController()
+
+    @classmethod
+    @TestAccountMoveSendCommon.setup_country('tw')
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.company_data['company'].write({
+            'l10n_tw_edi_ecpay_staging_mode': True,
+            'l10n_tw_edi_ecpay_merchant_id': '2000132',
+            'l10n_tw_edi_ecpay_hashkey': 'ejCk326UnaZWKisg',
+            'l10n_tw_edi_ecpay_hashIV': 'q9jcZX8Ib9LM8wYk',
+            'phone': '+886 123 456 781',
+        })
+        cls.partner_a.write({
+            'phone': '+886 123 456 789',
+            'contact_address': 'test address',
+            'company_type': 'person',
+        })
+
+        # We can reuse this invoice for the flow tests.
+        cls.basic_invoice = cls.init_invoice(
+            'out_invoice', partner=cls.partner_a, products=cls.product_a,
+        )
+        cls.basic_invoice.action_post()
+
+    def test_01_can_generate_file(self):
+        """
+        Simply test that with a valid configuration, we can generate the file.
+        """
+        with patch(CALL_API_METHOD, new=self._test_01_mock):
+            json_data = self.basic_invoice._l10n_tw_edi_generate_invoice_json()
+        self.assertTrue(json_data)
+
+        formatted_phone = self.basic_invoice._reformat_phone_number(self.basic_invoice.partner_id.phone)
+        # Validate the customer data
+        self.assertEqual(json_data.get("MerchantID"), self.company_data['company'].l10n_tw_edi_ecpay_merchant_id)
+        self.assertEqual(json_data.get("CustomerName"), self.basic_invoice.commercial_partner_id.name)
+        self.assertEqual(json_data.get("CustomerAddr"), self.basic_invoice.partner_id.contact_address)
+        self.assertEqual(json_data.get("CustomerEmail"), self.basic_invoice.partner_id.email)
+        self.assertEqual(json_data.get("CustomerPhone"), formatted_phone)
+        self.assertEqual(json_data.get("SalesAmount"), self.basic_invoice.amount_total)
```

Comment:
Try to use static values on the right side of the assert :thinking: 
The idea is that we expect specific name/address/... following a specific format/...
So we assert for these exact values